torch/csrc/autograd/functions/utils.cpp

```diff
@@ -9,14 +9,14 @@
 
 namespace torch { namespace autograd {
 
-variable_list wrap_outputs(const variable_list& inputs, tensor_list&& outputs,
+variable_list wrap_outputs(const variable_list& inputs, variable_list&& outputs,
```

Comment:
I don't understand what's going on here, can you explain?

This is only called for legacy_apply functions?  These should always be used with variable_data for some reason?

Reply:
This is called for `legacy_apply` functions and also `DelayedError::apply`. The main reason we need to change this from `tensor_list` to `variable_list` is because `outputs` are Variables now instead of Tensors (because we now use `.detach()` instead of `.tensor_data()` to build up the `outputs` list in `legacy_apply` and `DelayedError::apply`).

Reply:
I get that we had tensors and now we have variables, I'm trying to understand what the constraints of this code actually are.  So this is used with legacy apply, which we want to get rid of, right? (https://github.com/pytorch/pytorch/issues/16947).  How does legacy apply actually work?  The forwards is wrapped in a no-grad block?

Reply:
For legacy autograd functions, the forward is wrapped in a no-grad block https://github.com/pytorch/pytorch/blob/0c091380cc03b23e68dde7368f3b910c21deb010/torch/csrc/autograd/python_function.cpp#L646-L652 `legacy_apply` is only used in legacy autograd function's backward pass https://github.com/pytorch/pytorch/blob/0c091380cc03b23e68dde7368f3b910c21deb010/torch/csrc/autograd/python_function.cpp#L107-L118

In `legacy_apply`, we have to make a shallow-copy of the output from the backward function (via `detach` or `variable_data`) before calling `wrap_outputs`, because for a `backward` function that directly pass `grad_x` through:
```python
class PassthroughFunction(Function):
    ...
    def backward(self, grad_x):
        return grad_x
```
if `grad_x` requires grad, we will attach a new gradient edge to `grad_x` (via `autograd::create_gradient_edge(output, grad_fn)`), which will replace any previous gradient edge that `grad_x` has (which is an incorrect behavior). Hence we should make a shallow-copy of the output from the backward function in `legacy_apply` before calling `wrap_outputs`.